### PR TITLE
Adds basic command to validate package URLs

### PR DESCRIPTION
**Pull request by cameronscott137**
Per discussion in #54, this PR introduces a command to periodically check the status of package URLs.

## Summary
1. Every Sunday evening, the command grabs all packages that have not been tagged with "404 error." (This avoids us spamming contributors that have already been notified in step 4.) 
2. For each package, we check the response status code for both the `url` and `repo_url` attributes. If either returns anything other than a `200`, we do the following:
3. Tag the Package with a "404 error" tag, which indicates to users browsing NovaPackages that the package may no longer be maintained.
4. Send an email notification to the package author, and any associated contributors, inviting them to review their package and update the URL where necessary.

Here's an screenshot of the email notification that contributors will receive:

Screen Shot 2020-10-07 at 12 11 41 PM

## Tests
The feature is tested in `CheckPackageUrlsCommandTest`. Other tests in the suite are passing as expected.

## Questions/Feedback Needed
1. Is the `repo_url` attribute still in use? I see references to it in some places, but it looks like the `url` attribute is doing the heavy lifting these days. If so, I'll refactor to just validate `url` instead.  **Note:** running this command with seeder data will result in all packages getting marked as invalid, because `repo_url` is currently `null`. 
2. I'm not entirely sold on the tag language being "404 error." Something a bit more general feels appropriate, e.g. "URL error", but I'm struggling to identify the specific language, if folks have any stronger ideas.
3. Early on, I had attempted to validate the package's `composer_name` as well. But, rather than throw a 404, Packagist displays a search page, making it difficult for Zttp to parse a valid package from a missing one, which is a bummer. Ex: [https://packagist.org/packages/some-dev/package-name](https://packagist.org/packages/some-dev/package-name).

### CODE REVIEW

**Comment by marcusmoore** [comment]
Hey @cameronscott137 👋 

Sorry for the delay on responding to you on this. I'm planning on looking this over in detail on Friday 😄

**Comment by cameronscott137** [comment]
@marcusmoore totally fine! I'll take another look through before then and make sure I still agree with Cam from two weeks ago.

File: app/Notifications/NotifyContributorOfInvalidPackageUrl.php
Comment (marcusmoore) [inline app/Notifications/NotifyContributorOfInvalidPackageUrl.php:41]: This is good. I'm only leaving a comment here to remember that we'll need to update the text to prompt the user to re-enable the package if we end up using the `is_disabled` field.

**Review by marcusmoore — CHANGES_REQUESTED** [review]
Thanks again @cameronscott137!

😅  Lots of comments and questions but please don't feel overwhelmed. A lot of the changes requested are simple clean ups.

Below are my responses to the questions/feedback section:

> 1. Is the repo_url attribute still in use? I see references to it in some places, but it looks like the url attribute is doing the heavy lifting these days. If so, I'll refactor to just validate url instead. Note: running this command with seeder data will result in all packages getting marked as invalid, because repo_url is currently null.

Good catch...I did a quick look through and I think we can focus on the `url` attribute here. Do you agree?


> 2. I'm not entirely sold on the tag language being "404 error." Something a bit more general feels appropriate, e.g. "URL error", but I'm struggling to identify the specific language, if folks have any stronger ideas.

I'm in the same boat. Can we utilize the `is_disabled` attribute on the `Package` model and avoid showing unavailable packages on the site? The issue here would be that disabling a package doesn't allow users to come back and update that package since they'll see a 404...Do you have any thoughts on this?

> 3. Early on, I had attempted to validate the package's composer_name as well. But, rather than throw a 404, Packagist displays a search page, making it difficult for Zttp to parse a valid package from a missing one, which is a bummer. Ex: packagist.org/packages/some-dev/package-name.

That's unfortunate. I _think_, but I'm not sure, that we can just stick to checking urls right?

**Comment by cameronscott137** [comment]
@marcusmoore, No worries about the volume of comments—first time contributing to OSS, so appreciate your time and diligence. Notes below:

>  Lots of comments and questions but please don't feel overwhelmed. A lot of the changes requested are simple clean ups.

Do y'all have a recommended style guide I can follow for subsequent PRs? Would be happy to save you the trouble of highlighting silly formatting issues in the future.

> Good catch...I did a quick look through and I think we can focus on the url attribute here. Do you agree?

Yep, we're agreed.

> I'm in the same boat. Can we utilize the is_disabled attribute on the Package model and avoid showing unavailable packages on the site? The issue here would be that disabling a package doesn't allow users to come back and update that package since they'll see a 404...Do you have any thoughts on this?

That's interesting. The `is_disabled` attribute feels like a different level of escalation. Maybe something like this?

- The first time a 404 is detected, we send the author a notification, and tag the URL as "404 Error" (or whatever language we want to use.) In addition, we store the timestamp of the detected error.
- Using that timestamp as a basis, we can send a follow-up message to that author after two weeks.
- If, after a full month of errors, there's been no reconciliation, we disable the package, and send the author a unique URL to re-enable the package if they choose.

This approach would give maintainers ample time to easily address any URL issues, and help us keep the directory a little bit cleaner.

>That's unfortunate. I think, but I'm not sure, that we can just stick to checking urls right?

I think it's fine—the repo is likely going to be a more reliable source of truth for a package's activity.

**Comment by marcusmoore** [comment]
> No worries about the volume of comments—first time contributing to OSS, so appreciate your time and diligence

Awesome! That's great to hear 😄 

> Do y'all have a recommended style guide I can follow for subsequent PRs?

We use [tlint](https://github.com/tighten/tlint). The [Linters section](https://github.com/tighten/tlint#available-linters) is a good overview overview of what tlint looks for.

> That's interesting. The `is_disabled` attribute feels like a different level of escalation. Maybe something like this?

That's something I was feeling as well. I like what you are outlining but I don't necessarily want to show the user the `404 Error` tag...what do you think about adding a timestamp to the `Package` model called `marked_as_unavailable_at` were we populate that column if we get a 404 response? Then we can use it to send the follow-up email and disable the package after a month? In addition, we can use it as a flag to a warning message on the package's page.

File: app/Console/Commands/CheckPackageUrls.php
Comment (cameronscott137) [reply app/Console/Commands/CheckPackageUrls.php]: Yep, complete!

File: app/Notifications/NotifyContributorOfInvalidPackageUrl.php
Comment (cameronscott137) [reply app/Notifications/NotifyContributorOfInvalidPackageUrl.php]: Sure should—wasn't sure if you had a queue set up on production, so left it out. That's been corrected.

File: app/Jobs/CheckPackageUrls.php
Comment (cameronscott137) [reply app/Jobs/CheckPackageUrls.php]: Confirmed—now only notifying the package author, where available.

File: app/Jobs/CheckPackageUrls.php
Comment (cameronscott137) [reply app/Jobs/CheckPackageUrls.php]: You're right, that's a good call. I've modified this to check only for client errors: >= 400, and < 500. Any redirects, server issues, etc. will pass the test.

File: app/Jobs/CheckPackageUrls.php
Comment (cameronscott137) [reply app/Jobs/CheckPackageUrls.php]: Confirming that this has been removed—we're only checking the validity of the `url` attribute now.

**Comment by cameronscott137** [comment]
@marcusmoore Perfect—thanks for the tip on the Tlint! Will lint with abandon for all future work.

I like your suggestions. To confirm, the proposed flow is now:
- Check the `url` attribute. If invalid, log timestamp in new `marked_as_unavailable_at` attribute, and notify package author.
- Show visual indication of `marked_as_unavailable_at` somewhere within individual package UI.
- Any subsequent changes to the Package record (e.g. package author makes update) clears the `marked_as_unavailable_at` timestamp and resets the process.
- If Package is not updated, send follow up after two weeks.
- If, after a month, Package is not updated, mark as `is_disabled`, and send notification with unique URL to package author to re-enable.

We've removed any tagging, and instead rely on the two timestamp attributes for Package status.

Feels improved to me—does that seem reasonable to you?

**Comment by marcusmoore** [comment]
Glad you like tlint 😄 It's a great tool (that I had no part of creating).

This new flow looks great. A couple things pop out:
> Any subsequent changes to the Package record

I think we shouldn't clear `marked_as_unavailable_at` unless they updated the `url` attribute. This will avoid users changing the name (for example) and the app thinking that the package is back to a good state.

> send notification with unique URL to package author to re-enable

Maybe we don't need to send them a unique url...We can probably adapt the guard in the [`RouteServiceProvider`](https://github.com/tighten/novapackages/blob/master/app/Providers/RouteServiceProvider.php#L41-L46) to let package authors throw to the edit page...I'll leave that up to you and whatever you think is best to implement.

Thanks again for your hard work on this! I'm going to sign off for the weekend soon so I might not get back to this until late next week but I'm going to throw the `hacktoberfest-accepted` label on it so, if it's not merge by the 31st, you still get Hacktoberfest credit 😄

**Comment by cameronscott137** [comment]
@marcusmoore sorry for the delay. Here's an updated flow:

- A weekly command checks the package `url` attribute. If invalid, the command logs a timestamp in new `marked_as_unavailable_at` attribute, and notifies the package author.
- If a package is marked as unavailable, we display a notice on the public package page.
- If the package `url` is updated at any point (by any user type who has edit access), the app resets the package to available. 
- If the package `url` is still not updated after two weeks, we send a follow-up notification to the author.
- If, after a month, the package `url` is not updated, the app marks the package as `is_disabled`. At that point, a package author can still sign in, access the edit view, update the Package URL, and reset the package's availability.

Here's what the public package view for an unavailable package looks like: 
Screen Shot 2020-10-29 at 7 00 50 PM

Here's the email that package authors receive after two weeks: 
Screen Shot 2020-10-30 at 12 08 45 PM

Here's the notice that appears on the edit form for an invalid package:
Screen Shot 2020-10-30 at 10 09 41 AM

Here's the email that package authors receive after the package has been disabled:
Screen Shot 2020-10-30 at 12 07 57 PM


This turned into a bit more of an in-depth PR than I'd intended, but see how this feels, and let me know if you spot any additional gaps in logic or functionality.

(As an aside, it looks like a build process might be failing due to the Laravel 8 update. I'll try to look into that this afternoon (I think it's model factory syntax), but it shouldn't impact you reviewing the feature in the meantime.)

**Review by marcusmoore — APPROVED** [review]
Tons of work went into this PR @cameronscott137 and it looks great! Thanks for your hard work and willingness to push through a big PR 😄

**Comment by gregoriohc** [comment]
Hi @cameronscott137 and @marcusmoore!

I'm not sure if the problem is with the code of this PR or maybe a temporary server issue, but I have received the "invalid package" email for each of my packages, and the URLs of them are marked as inaccesible. I have double checked them and the URLs are still working (they are Github URLs).
Can you have a look at it?

Thanks!

**Comment by cameronscott137** [comment]
Thanks @gregoriohc, I'd noticed that, too. @marcusmoore, it looks like we [had](https://novapackages.com/packages/laraning/nova-time-field) [some](https://novapackages.com/packages/optimistdigital/nova-sortable) [false](https://novapackages.com/packages/optimistdigital/nova-multiselect-filter) positives in the first run of the new command 😬😬😬.

I apologize—unsure whether it's an issue with the PR, or something with production data or environment? Going to dig into this more today.

**Comment by cameronscott137** [comment]
@marcusmoore In digging through this further, I'm 95% sure we hit a rate-limit on Github, which would return a 403 error, and falls right in line with the `clientError()` range that we're checking for. Again, that's my bad—sorry for the inconvenience.

**Comment by marcusmoore** [comment]
@gregoriohc I agree and think we hit rate-limiting too.

It happens. Let's work together on getting it fixed up in the next few days 😄

**Comment by marcusmoore** [comment]
Just realized I tagged the wrong person...Monday is going great so far!

**Comment by cameronscott137** [comment]
@marcusmoore Hey, here's hoping that, come Friday, we can look back and decide this was the worst thing that happened to us this week.

When you have time, I think can solve this without too much change. Two core issues: 

- Before notifying package authors, we do not allow any buffer to account for third-party service outages and other temporary issues.
- As written, we run the risk of breaching third-party rate-limits.

To resolve, we modify the existing flow to:
- Run existing command daily, and continue to check URL statuses. If a URL gets flagged as invalid, we store the timestamp. Conversely, if a URL resolves correctly, we make sure the timestamp is set to `null`.
- To avoid rate-limiting, we can `sleep()` for a few seconds after queuing each job.
- If the timestamp for a URL had been previously stored, we check it against a notification schedule. If that timestamp is 7 days old—and therefore, has failed the check 7 times—we send our first notification. If it fails for 14 days straight, we send our second notification. It it fails every day for a full month, we disable and send our final notification.

This gives the system *plenty* of buffer to weed out temporary outages and other issues.

**Comment by marcusmoore** [comment]
> Hey, here's hoping that, come Friday, we can look back and decide this was the worst thing that happened to us this week.

Good point 😅 

Overall, I think your proposal makes sense. I'm going to create a new issue in the next week detailing thoughts I have and opening up the conversation. I'll tag you so we can continue chatting there.

Have a great weekend 😄

**Pull request merged**
